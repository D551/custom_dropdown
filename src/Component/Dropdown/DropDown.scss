.dropdown-container {
  display: flex;
  flex-direction: column;
  margin: auto;
  width: 30%;

  .inputWrap {
    margin: 8px 0;
    border: 1px solid #ccc;
    border-radius: 4px;
    box-sizing: border-box;
    position: relative;
  
    .inputBox {
      width: 100%;
      padding: 12px 25px 12px 20px;
      border: 0;
      border-radius: 4px;
      box-sizing: border-box;
    }

    .arrow {
      border: solid black;
      border-width: 0 3px 3px 0;
      display: inline-block;
      padding: 3px;
      position: absolute;
      right: 10px;
    }

    .up {
      transform: rotate(-135deg);
      -webkit-transform: rotate(-135deg);
      top: 16px;
    }

    .down {
      transform: rotate(45deg);
      -webkit-transform: rotate(45deg);
      top: 13px;
    }
  }

  .list-container {
    max-height: 10em;
    display: flex;
    flex-direction: column;
    overflow: auto;
    display: inline-block;
    border: 1px solid;
    border-radius:5px;

    .list-card {
      width: 100%;
      height: 80px;
      padding: 1rem;
      position: relative;
      display: flex;
      justify-content: center;
      box-sizing: border-box;
      align-content: center;
      align-items: center;
      background: white;
      border: 0px;
      border-top: 1px solid;
      cursor: pointer;


      .name {
        font-size: medium;
        font-weight: 600;

      }

      .mail {
        padding-top: 0.3em;
      }
    }
    .list-card:first-child {
      border-top: 0px;
    }

    .color-icon {
      width: 24px;
      height: 24px;
      padding: 0.5rem;
      text-align: center;
      border-radius: 8px;
      color: white;
      line-height: 24px;
      font-weight: 600;
      pointer-events: none;
    }

    span.users-info {
      width: calc(100% - 52px);
      text-align: left;
      margin-left: 10px;
    }

    .list-card:hover {
      background: #e7fafa;
    }

    .list-card:hover span.users-info {
      color: rgb(84, 168, 140);
      pointer-events: none;
    }
    
  }
}