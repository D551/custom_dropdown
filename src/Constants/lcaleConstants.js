export const Constants={
    PLACEHOLDER:'Choose Manager',
    NO_USER_FOUND:'No user found!!'
}