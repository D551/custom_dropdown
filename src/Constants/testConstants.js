
export const MOCK_DROPDOWN_LIST = [
    {
        id: '323',
        firstName: 'Harriet',
        lastName: 'McKinney',
        name: 'Harriet McKinney',
        accountId: '324',
        profileColor: '#97802D',
        email: 'harriet.mckinney@kinetar.com',
    },
    {
        id: '139',
        firstName: 'Harriet',
        lastName: 'Banks',
        name: 'Harriet Banks',
        accountId: '140',
        profileColor: '#AE3093',
        email: 'harriet.banks@kinetar.com',
    },
    {
        id: '142',
        firstName: 'Mathilda',
        lastName: 'Summers',
        name: 'Mathilda Summers',
        accountId: '143',
        profileColor: '#13ACFC',
        email: 'mathilda.summers@kinetar.com',
    },
];

export const MOCK_API_REPONSE = {
    data: [
        {
            type: 'employees',
            id: '323',
            links: {
                self: 'http://localhost:3000/v1/employees/323',
            },
            attributes: {
                identifier: null,
                firstName: 'Harriet',
                lastName: 'McKinney',
                name: 'Harriet McKinney',
                features: ['engagement'],
                avatar: null,
                employmentStart: '2016-01-31T00:00:00.000Z',
                external: false,
                'Last Year Bonus': 3767,
                'Business Unit': 'Sales',
                'Commute Time': 34,
                Age: '1984-02-08',
                Department: 'Customer Care',
                Gender: 'Female',
                'Job Level': 'Manager',
                'Local Office': 'Kuala Lumpur',
                '% of target': 88,
                Region: 'APAC',
                Salary: 76000,
                Tenure: '2014-05-31',
            },
            relationships: {
                company: {
                    data: {
                        type: 'companies',
                        id: '5',
                    },
                },
                account: {
                    data: {
                        type: 'accounts',
                        id: '324',
                    },
                },
                phones: {
                    data: [],
                },
                Manager: {
                    data: {
                        type: 'employees',
                        id: '201',
                    },
                },
            },
        },
        {
            type: 'employees',
            id: '139',
            links: {
                self: 'http://localhost:3000/v1/employees/139',
            },
            attributes: {
                identifier: null,
                firstName: 'Harriet',
                lastName: 'Banks',
                name: 'Harriet Banks',
                features: ['engagement'],
                avatar: null,
                employmentStart: '2016-01-31T00:00:00.000Z',
                external: false,
                'Last Year Bonus': 215835,
                'Business Unit': 'People & Operations',
                'Commute Time': 18,
                Age: '1984-04-19',
                Department: 'Learning and Development',
                Gender: 'Female',
                'Job Level': 'Executive',
                'Local Office': 'New York',
                '% of target': 12,
                Region: 'AMER',
                Salary: 332000,
                Tenure: '2009-06-30',
            },
            relationships: {
                company: {
                    data: {
                        type: 'companies',
                        id: '5',
                    },
                },
                account: {
                    data: {
                        type: 'accounts',
                        id: '140',
                    },
                },
                phones: {
                    data: [],
                },
            },
        },
        {
            type: 'employees',
            id: '142',
            links: {
                self: 'http://localhost:3000/v1/employees/142',
            },
            attributes: {
                identifier: null,
                firstName: 'Mathilda',
                lastName: 'Summers',
                name: 'Mathilda Summers',
                features: ['engagement'],
                avatar: null,
                employmentStart: '2016-01-31T00:00:00.000Z',
                external: false,
                'Last Year Bonus': 95050,
                'Business Unit': 'Marketing',
                'Commute Time': 131,
                Age: '1976-12-19',
                Department: 'Media',
                Gender: 'Female',
                'Job Level': 'Executive',
                'Local Office': 'London',
                '% of target': 166,
                Region: 'EMEA',
                Salary: 248000,
                Tenure: '2000-05-15',
            },
            relationships: {
                company: {
                    data: {
                        type: 'companies',
                        id: '5',
                    },
                },
                account: {
                    data: {
                        type: 'accounts',
                        id: '143',
                    },
                },
                phones: {
                    data: [],
                },
                Manager: {
                    data: {
                        type: 'employees',
                        id: '139',
                    },
                },
            },
        }
    ],
    included: [
        {
            type: 'accounts',
            id: '140',
            links: {
                self: 'http://localhost:3000/v1/accounts/140',
            },
            attributes: {
                email: 'harriet.banks@kinetar.com',
                locale: null,
                timezone: null,
                bouncedAt: null,
                bounceReason: null,
                localeEffective: 'en-GB',
                timezoneEffective: null,
            },
        },
        {
            type: 'accounts',
            id: '143',
            links: {
                self: 'http://localhost:3000/v1/accounts/143',
            },
            attributes: {
                email: 'mathilda.summers@kinetar.com',
                locale: null,
                timezone: null,
                bouncedAt: null,
                bounceReason: null,
                localeEffective: null,
                timezoneEffective: null,
            },
        },
        {
            type: 'accounts',
            id: '324',
            links: {
                self: 'http://localhost:3000/v1/accounts/324',
            },
            attributes: {
                email: 'harriet.mckinney@kinetar.com',
                locale: null,
                timezone: null,
                bouncedAt: null,
                bounceReason: null,
                localeEffective: null,
                timezoneEffective: null,
            },
        }
    ],
    meta: {
        page: {
            total: 9,
        },
    },
    links: {
        self: 'http://localhost:3000/v1/employees/contexts/company_5?per_page=25&q=harriet&sort=score&order=desc&include=account%2Cphones%2CLast%20Year%20Bonus%2CBusiness%20Unit%2CCommute%20Time%2CManager&simple=false',
        first:
            'http://localhost:3000/v1/employees/contexts/company_5?per_page=25&q=harriet&sort=score&order=desc&include=account%2Cphones%2CLast%20Year%20Bonus%2CBusiness%20Unit%2CCommute%20Time%2CManager&simple=false',
        last: 'http://localhost:3000/v1/employees/contexts/company_5?per_page=25&page=1&q=harriet&sort=score&order=desc&include=account%2Cphones%2CLast%20Year%20Bonus%2CBusiness%20Unit%2CCommute%20Time%2CManager&simple=false',
    },
};
