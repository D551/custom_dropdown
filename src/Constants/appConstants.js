import { Constants } from "./lcaleConstants";

export const NO_USER_FOUND_OBJECT = {
    email: "",
    id: "0",
    name: Constants.NO_USER_FOUND,
    profileColor: "rgb(220 221 227)",
    notClickable: true
}