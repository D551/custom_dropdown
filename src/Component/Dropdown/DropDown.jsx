import React from 'react';
import PropTypes from 'prop-types';
import { NO_USER_FOUND_OBJECT } from '../../Constants/appConstants';

import './DropDown.scss';

const ManagersDropDown = (
    {
        options,
        placeholder,
        onChange
    }
) => {
    const [_options, _setOptions] = React.useState(options);
    const [showDropdown, setShowDropdown] = React.useState(false);
    const [textValue, setTextValue] = React.useState('');

    const getSeletedListText = (e) => {
        setTextValue(e.target.value);
        setShowDropdown(false);
        const selectedOption = options.filter((x) => { return x.id === e.target.id })
        onChange(selectedOption[0]);
    }

    const createOptionsList = (info) => {
        return (
            <button
                className='list-card'
                key={info.id}
                onClick={!info.notClickable ? getSeletedListText : () => { }}
                value={info.name}
                id={info.id}
            >
                {
                    info.firstName && info.lastName &&
                    <span className="color-icon" style={{ background: info.profileColor }}>
                        {info.firstName[0] + info.lastName[0]}
                    </span>
                }
                <span className="users-info">
                    <div className='name'>{info.name}</div>
                    <div className="mail">{info.email}</div>
                </span>
            </button>
        )
    }

    const filterList = (options, searchTest) => {
        const filteredList = options.filter((option) => {
            return option.name.toLowerCase().replaceAll(/\s/g, '')
                .includes(searchTest.toLowerCase().replaceAll(/\s/g, ''))
        })
        return filteredList
    }

    const onTextChange = (e) => {
        setTextValue(e.target.value)
        if (e.target.value.length) {
            const searchedlist = filterList(options, e.target.value.toLowerCase());
            if (searchedlist.length) {
                _setOptions(searchedlist)
            } else {
                _setOptions([NO_USER_FOUND_OBJECT])
            }
        } else {
            _setOptions(options)
        }
    }

    const onTextBoxFocus = () => {
        if (textValue.length) {
            _setOptions(filterList(options, textValue.toLowerCase()))
        }
        setShowDropdown(true)
    }

    const handleBlur = (e) => {
        if (!e.currentTarget.contains(e.relatedTarget)) {
            setShowDropdown(false);
        }
    };

    return (
        <div className='dropdown-container'
            onBlur={handleBlur}>
            <div className="inputWrap">
                <input type="text"
                    className='inputBox'
                    onFocus={onTextBoxFocus}
                    onChange={onTextChange}
                    value={textValue}
                    placeholder={placeholder}
                    data-testid="input-box"
                />
                <i className={showDropdown ? 'arrow up' : 'arrow down'}
                    onClick={() => { setShowDropdown(!showDropdown) }}
                />
            </div >
            {
                showDropdown &&
                <div className='list-container' data-testid="list-container">
                    {
                        _options?.map((manager) => {
                            return createOptionsList(manager)
                        })
                    }
                </div>
            }
        </div >
    )
};

ManagersDropDown.propTypes = {
    options: PropTypes.array,
    placeholder: PropTypes.string,
    onChange: PropTypes.func,
};

ManagersDropDown.defaultProps = {
    options: [],
    placeholder: '',
    onChange: () => { },
};

export default ManagersDropDown;
