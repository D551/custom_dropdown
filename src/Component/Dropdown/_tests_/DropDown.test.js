/* eslint-disable testing-library/prefer-screen-queries */
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import DropDown from '../DropDown';
import { MOCK_DROPDOWN_LIST } from '../../../Constants/testConstants'
import { NO_USER_FOUND_OBJECT } from '../../../Constants/appConstants'

describe('should test the DropDown component', () => {
    const mockFunction = jest.fn();
    const mockProps = {
        options: MOCK_DROPDOWN_LIST,
        placeholder: '',
        onChange: () => { }
    }

    const setup = () => {
        const utils = render(< DropDown {...mockProps} />)
        const input = utils.getByTestId('input-box')
        return {
            input,
            ...utils,
        }
    }
    test('should render dropDown', () => {
        const { asFragment } = render(< DropDown {...mockProps} />);
        expect(asFragment(< DropDown {...mockProps} />)).toMatchSnapshot();
    });

    test('should render list onFocus of input box', () => {
        const { getByTestId } = render(< DropDown {...mockProps} />);
        fireEvent.focusIn(getByTestId('input-box'));
        expect(getByTestId('list-container')).toBeInTheDocument();
    });

    test('should render list onChnage of input box', () => {
        const { input, getByTestId } = setup();
        fireEvent.focusIn(getByTestId('input-box'));
        fireEvent.change(input, { target: { value: 'tMc' } });
        expect(input.value).toBe('tMc');
        expect(getByTestId('list-container')).toBeInTheDocument();
    });

    test('should hide the dropdown when clicked outside', () => {
        jest.spyOn(React, 'useState').mockReturnValue([false, mockFunction]);
        const { getByTestId } = setup();
        fireEvent.focusIn(getByTestId('input-box'));
        expect(mockFunction).toHaveBeenCalledWith(true);
        fireEvent.focusOut(getByTestId('input-box'));
        expect(mockFunction).toHaveBeenCalledWith(false);
    });

    test('Dropdown conatines 2 child when matching text is typed', () => {
        jest.spyOn(React, 'useState').mockReturnValue([false, mockFunction]);
        const { input, getByTestId } = setup();
        fireEvent.focusIn(getByTestId('input-box'));
        fireEvent.change(input, { target: { value: 'ha' } });
        expect(mockFunction).toHaveBeenCalledWith([MOCK_DROPDOWN_LIST[0], MOCK_DROPDOWN_LIST[1]]);
    });

    test('Should show no user found', () => {
        jest.spyOn(React, 'useState').mockReturnValue([false, mockFunction]);
        const { input, getByTestId } = setup();
        fireEvent.focusIn(getByTestId('input-box'));
        fireEvent.change(input, { target: { value: 'xxxx' } });
        expect(mockFunction).toHaveBeenCalledWith([NO_USER_FOUND_OBJECT]);
    });
});
