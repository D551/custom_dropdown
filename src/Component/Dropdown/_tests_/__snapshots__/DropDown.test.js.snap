// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test the DropDown component should render dropDown 1`] = `
<DocumentFragment>
  <div
    class="dropdown-container"
  >
    <div
      class="inputWrap"
    >
      <input
        class="inputBox"
        data-testid="input-box"
        placeholder=""
        type="text"
        value=""
      />
      <i
        class="arrow down"
      />
    </div>
  </div>
</DocumentFragment>
`;
